crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -165,6 +165,18 @@ impl CheckpointExecutor {
             .checkpoint_store
             .get_highest_executed_checkpoint()
             .unwrap();
+
+        if let Some(highest_executed) = &highest_executed {
+            if epoch_store.epoch() == highest_executed.epoch()
+                && highest_executed.is_last_checkpoint_of_epoch()
```

Comment:
This is a bit confusing to me, so perhaps we could do better with comments in a few other places.
Is the variable `highest_executed` no longer mean the value of the highest executed watermark, due to the last checkpoint of epoch? i.e. when we reach line 200 on the call check_epoch_last_checkpoint, `highest_executed` may be the last checkpoint, but we in fact hasn't bumped the watermark yet?

Reply:
Yes, you are correct that that is the case when we are dealing with the final checkpoint, if there were no restarts. When we enter check_epoch_last_checkpoint, highest_executed may be the final checkpoint, but the watermark hasn't been bumped yet. I don't really know how to improve this without doing a much bigger refactoring.

I will add some comments